File: C exercices/credit.c
#include <cs50.h>
#include <stdio.h>

int valido(long number);
string brand(long number);
int main(void)
{
    //reciving a valid number
    long number = 0;
    do
    {
        number = get_long("Number: ");
    }
    while (number <= 0);

    //verifying if the card is valid
    int validador = valido(number);
    
    //finding out the card brand
    if (validador % 10 == 0)
    {
        string card_brand = brand(number);
        printf("%s\n", card_brand);
    }
    else
    {
        printf("INVALID\n");
    }
}

// function that verify if the number is a valid card
int valido(long number)
{
    //Conjunto de variaveis
    long resultado1 = 0, resultado2 = 0;
    int sobra = 0, validador1 = 0, validador2 = 0;
    // Loop
    do
    {
        validador2 += (number % 10);
        resultado1 = number / 10;
        resultado2 = resultado1 / 10;
        sobra = 2 * (resultado1 % 10);
        //desvio
        if (sobra < 10)
        {
            validador1 += sobra;
        }
        else
        {
            validador1 += 1 + (sobra - 10);
        }
        number = resultado2;
    }
    while (number > 0);
    int validador = validador1 + validador2;
    return validador;
}

// function that defines the credit card brand
string brand(long number)
{
    //contador
    long i = 0;
    int cont = 1;
    for (i = number; i > 100;)
    {
        i /= 10;
        cont++;
    }
    //Mastercards
    if (i == 51 || i == 52)
    {
        return "MASTERCARD";
    }
    else if (i == 53 || i == 54)
    {
        return "MASTERCARD";
    }
    else if (i == 55)
    {
        return "MASTERCARD";
    }
    //AMEX
    else if (i == 34 || i == 37)
    {
        return "AMEX";
    }
    //VISA
    else if (cont + 1 == 13 && i / 10 == 4)
    {
        return "VISA";
    }
    else if (cont + 1 == 16 && i / 10 == 4)
    {
        return "VISA";
    }
    //INVALID
    else
    {
        return "INVALID";
    }
}

File: problems_set/pset6/credit.py
def main():
    while True:
        number = int(input("Number: "))
        if number <= 0:
            break
    
    is_valid = valid(number)

    if (is_valid % 10 == 0):
        card_brand = brand(number)
        print(card_brand)
    else:
        print("INVALID")


def valid(number):
    particial1 = parcial2 = result1 = result2 = remainder = 0

    while True:
        parcial2 += (number % 10)
        result1 = number / 10
        result2 = result1 / 10
        remainder = 2 * (result1 % 10)

        if (remainder < 10):
            particial1 += remainder
        else:
            particial1 += 1 + (remainder - 10)
        number = result2

        if (number > 0):
            break

        return particial1 + parcial2


def brand(number):
    first_numbers = number
    total_number = 1

    for c in range(number, 100):
        first_numbers /= 10
        total_number += 1
    
    if (51 <= first_numbers <= 55 ):
        return 'MASTERCARD'
    elif (first_numbers == 34 or first_numbers == 37):
        return 'AMEX'
    elif (first_numbers / 10 == 4 and total_number + 1 == 13 or first_numbers / 4 == 10 and total_number + 1 == 16) :
        return 'Visa'
    else:
        return 'INVALID'


main()